2016/submission/ref-util.pod

```diff
@@ -18,49 +30,101 @@ maintain: Checking the validity of references using the C<ref> function.
+
+Since under strict a typo in C<is_plain_hash> would cause a compile time error
+any fat-fingered elf problems would be caught. However, that's just introduced
+another subroutine call, and since these ref checks are used all over Santa's
+code base, the cost of doing that adds up!
+
+The second problem with the pattern was the fact that it was strictly doing more
+work that it needed to.  Consider what the following was actually doing:
+
+    ref $args eq 'HASH'
+
+Under the hood perl is inspecting bits on the reference to see if it's a C<HV>
+(a C<hash value>), then building a standard SV string containing the four
+characters C<HASH>, then in a separate operation comparing that string character
+by character with the C<HASH> string was passed in in the code.  That, of
+course, all happens really quickly, but it's more work (and more Perl virtual
+ops) than need happen just to tell if this was a hashref or not!
+
+So the code review had found two problems:  One, an abstraction was needed to
+make it more reliable.  Two, the code needed to involve less stuff to make it
+faster.
+
+=head2 The Voice of an Angel
+
+Thank goodness the Wise Old Elf read a L<CPAN Weekly|http://cpan-weekly.org/>
+email about this very pattern! Because of this he decided to try out a new
+module, L<Ref::Util>.
+
+L<Ref::Util> provides a set of helpful functions to determine what kind of
+reference a variable is. It abstracts several awkward ref checking patterns into
+new keywords.  So, as our above example:
 
     ref $args eq 'HASH'
         or croak('Sled expects arguments as a hashref');
 
-    # after
+Can be written as:
+
     use Ref::Util qw<is_plain_hash>;
 
     is_plain_hash($args)
         or croak('Sled expects arguments as a hashref');
 
 Or maybe we want to check for a blessed array reference, making sure it
-isn't accidentally blessed:
+isn't accidentally blessed? The following:
 
-    use Carp qw<croak>;
-
-    # before
     use Scalar::Util qw<blessed reftype>;
 
     blessed $args && ref $args && reftype($args) eq 'ARRAY'
         or Carp::croak('Uh oh, we require a blessed reference');
 
-    # after
+Can much more succinctly be written with the new C<is_blessed_arrayref> keyword:
+
     use Ref::Util qw<is_blessed_arrayref>;
 
     is_blessed_arrayref($args)
         or Carp::croak('Uh oh, we require a blessed reference');
 
 Not writing a lot of code but still being accurate is something Santa
-appreciates, but that's not the best part.
+appreciates, but that's not the best part....
+
+The Ref::Util module introduces new B<keywords> not just new I<functions> using
+Perl's pluggable keyword infrastructure.  The guts of the checks are implemented
+in fast C code (and skipping all the unnecessary string construction) - and
+since the function calls in our code get compiled using custom op codes into a
+single op in the tree means it's just as fast as the normal Perl inbuilt
+functions without the subroutine calling overhead!
```

Comment:
Sort of. It's correct it gets compiled using custom op codes into a single op. That's the long explanation I gave above. (Sorry for the repeat.)

It is in fact **faster** since it's fewer ops, but even the XS is faster, I believe, because it does the comparison in the XS function (and still fewer ops).